strawberryfields/configuration.py

```diff
@@ -0,0 +1,189 @@
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Configuration
+=============
+
+**Module name:** :mod:`strawberryfields.configuration`
+
+.. currentmodule:: strawberryfields.configuration
+
+This module contains the :class:`Configuration` class, which is used to
+load, store, save, and modify configuration options for Strawberry Fields.
+
+Behaviour
+---------
+
+On first import, Strawberry Fields attempts to load the configuration file `config.toml`, by
+scanning the following three directories in order of preference:
+
+1. The current directory
+2. The path stored in the environment variable ``SF_CONF``
+3. The default user configuration directory:
+
+   * On Linux: ``~/.config/strawberryfields``
+   * On Windows: ``~C:\Users\USERNAME\AppData\Local\Xanadu\strawberryfields``
+   * On MacOS: ``~/Library/Preferences/strawberryfields``
+
+If no configuration file is found, a warning message will be displayed in the logs,
+and all device parameters will need to be passed as keyword arguments when
+loading the device.
+
+The user can access the initialized configuration via `strawberryfields.config`, view the
+loaded configuration filepath, print the configurations options, access and modify
+them via keys, and save/load new configuration files.
+
+Configuration files
+-------------------
+
+The configuration file `config.toml` uses the `TOML standard <https://github.com/toml-lang/toml>`_,
+and has the following format:
+
+.. code-block:: toml
+
+    [API]
+    # Options for the Strawberry Fields Cloud API
+    authentatication_token = "071cdcce-9241-4965-93af-4a4dbc739135"
+    hostname = "localhost"
+    use_ssl = true
+
+Summary of options
+------------------
+
+todo
+
+Summary of methods
+------------------
+
+.. currentmodule:: strawberryfields.configuration.Configuration
+
+.. autosummary::
+    path
+    load
+    save
+
+Code details
+~~~~~~~~~~~~
+
+.. currentmodule:: strawberryfields.configuration
+
+"""
+import os
+import logging as log
+
+import toml
+from appdirs import user_config_dir
+
+log.getLogger()
+
+
+DEFAULT_CONFIG = {"api": {"authentatication_token": "", "hostname": "localhost", "use_ssl": True}}
+
+
+class ConfigurationError(Exception):
+    """Exception used for configuration errors"""
+
+
+class Configuration:
+    """Configuration class.
+
+    This class is responsible for loading, saving, and storing StrawberryFields
+    and plugin/device configurations.
+
+    Args:
+        name (str): filename of the configuration file.
+        This should be a valid TOML file. You may also pass an absolute
+        or a relative file path to the configuration file.
+    """
+
+    def __str__(self):
+        return "{}".format(self._config)
```

Comment:
I've been using f-strings lately and find them more readable -- e.g. `return f'{self._config}'`.

Reply:
Unfortunately we still have SF targeting Python 3.5, which means no f strings for now (as nice as they are)